web3_proxy/src/frontend/users.rs

```diff
@@ -853,3 +855,135 @@ pub async fn user_stats_detailed_get(
 
     Ok(response)
 }
+
+/// `GET /user/stats/detailed` -- Use a bearer token to get the user's key stats such as bandwidth used and methods requested.
+///
+/// If no bearer is provided, detailed stats for all users will be shown.
+/// View a single user with `?user_id=$x`.
+/// View a single chain with `?chain_id=$x`.
+///
+/// Set `$x` to zero to see all.
+///
+/// TODO: this will change as we add better support for secondary users.
+#[debug_handler]
+pub async fn admin_change_user_roles(
+    Extension(app): Extension<Arc<Web3ProxyApp>>,
+    bearer: Option<TypedHeader<Authorization<Bearer>>>,
+    Query(params): Query<HashMap<String, String>>,
+) -> FrontendResult {
+
+    // Make sure that the bearer exists, and has admin rights ...
+    let user_address: Vec<u8> = params
+        .get("user_address")
+        .ok_or_else(||
+            FrontendErrorResponse::StatusCode(
+                StatusCode::BAD_REQUEST,
+                "Unable to find user_address key in request".to_string(),
+                None,
+            )
+        )?
+        .parse::<Address>()
+        .map_err(|err| {
+            FrontendErrorResponse::StatusCode(
+                StatusCode::BAD_REQUEST,
+                "Unable to parse user_address as an Address".to_string(),
+                Some(err.into()),
+            )
+        })?
+        .to_fixed_bytes().into();
+    let user_tier_title = params
+        .get("user_tier_title")
+        .ok_or_else(|| FrontendErrorResponse::StatusCode(
+            StatusCode::BAD_REQUEST,
+            "Unable to get the user_tier_title key from the request".to_string(),
+            None,
+        ))?;
+
+    // Create database connections and all that
+    let db_conn = app.db_conn().context("admin_change_user_roles needs a db")?;
+    let db_replica = app
+        .db_replica()
+        .context("admin_change_user_roles needs a db replica")?;
+    let mut redis_conn = app
+        .redis_conn()
+        .await
+        .context("admin_change_user_roles had a redis connection error")?
+        .context("admin_change_user_roles needs a redis")?;
+
+    // TODO: Make a single query, where you retrieve the user, and directly from it the secondary user (otherwise we do two jumpy, which is unnecessary)
+    // get the user id first. if it is 0, we should use a cache on the app
+    let user_id = get_user_id_from_params(&mut redis_conn, &db_conn, &db_replica, bearer, &params).await?;
+
+    let mut response_body = HashMap::new();
+    response_body.insert(
+        "user_id",
+        serde_json::Value::Number(user_id.into()),
+    );
+
+    // Get both the user and user role
+    let user: user::Model = user::Entity::find_by_id(user_id)
+        .one(&db_conn)
```

Comment:
this should use `db_replica`

Reply:
not entirely sure, but what is the (practical) difference to why we have both a db_replica, and a db_conn? speed?

Reply:
db_replica is read only and replicated all over the world. It's much faster to query since everything doesn't have to go all the way to Ohio. 

If you are just reading the data, use replica.

If you are reading the data to modify and then write it or are writing brand new data, use the main db.